pkg/util/metrics_helper.go

```diff
@@ -0,0 +1,311 @@
+		}
+
+		perMetricName[name] = m
+	}
+
+	d[userID] = perMetricName
+	return nil
+}
+
+func (d MetricFamiliesPerUser) SendSumOfCounters(out chan<- prometheus.Metric, desc *prometheus.Desc, counter string) {
+	result := float64(0)
+	for _, perMetric := range d {
+		result += sum(perMetric[counter], counterValue)
+	}
+
+	out <- prometheus.MustNewConstMetric(desc, prometheus.CounterValue, result)
+}
+
+func (d MetricFamiliesPerUser) SendSumOfCountersWithLabels(out chan<- prometheus.Metric, desc *prometheus.Desc, counter string, labelNames ...string) {
+	type counterResult struct {
+		value       float64
+		labelValues []string
+	}
+
+	result := map[string]counterResult{}
+
+	for _, userMetrics := range d { // for each user
+		metricsPerLabelValue := getMetricsWithLabelNames(userMetrics[counter], labelNames)
+
+		for key, mlv := range metricsPerLabelValue {
+			for _, m := range mlv.metrics {
+				r := result[key]
+				if r.labelValues == nil {
+					r.labelValues = mlv.labelValues
+				}
+
+				r.value += m.GetCounter().GetValue()
+				result[key] = r
+			}
+		}
+	}
+
+	for _, cr := range result {
+		out <- prometheus.MustNewConstMetric(desc, prometheus.CounterValue, cr.value, cr.labelValues...)
+	}
+}
+
+func (d MetricFamiliesPerUser) SendSumOfCountersPerUser(out chan<- prometheus.Metric, desc *prometheus.Desc, counter string) {
+	for user, perMetric := range d {
+		v := sum(perMetric[counter], counterValue)
+
+		out <- prometheus.MustNewConstMetric(desc, prometheus.CounterValue, v, user)
+	}
+}
+
+func (d MetricFamiliesPerUser) SendSumOfGauges(out chan<- prometheus.Metric, desc *prometheus.Desc, gauge string) {
+	result := float64(0)
+	for _, perMetric := range d {
+		result += sum(perMetric[gauge], gaugeValue)
+	}
+	out <- prometheus.MustNewConstMetric(desc, prometheus.GaugeValue, result)
+}
+
+func (d MetricFamiliesPerUser) SendSumOfGaugesWithLabels(out chan<- prometheus.Metric, desc *prometheus.Desc, counter string, labelNames ...string) {
+	type gaugeResult struct {
+		value       float64
+		labelValues []string
+	}
+
+	result := map[string]gaugeResult{}
+
+	for _, userMetrics := range d { // for each user
+		metricsPerLabelValue := getMetricsWithLabelNames(userMetrics[counter], labelNames)
+
+		for key, mlv := range metricsPerLabelValue {
+			for _, m := range mlv.metrics {
+				r := result[key]
+				if r.labelValues == nil {
+					r.labelValues = mlv.labelValues
+				}
+
+				r.value += m.GetGauge().GetValue()
+				result[key] = r
+			}
+		}
+	}
+
+	for _, cr := range result {
+		out <- prometheus.MustNewConstMetric(desc, prometheus.CounterValue, cr.value, cr.labelValues...)
+	}
+}
+
+func (d MetricFamiliesPerUser) SendSumOfSummaries(out chan<- prometheus.Metric, desc *prometheus.Desc, summaryName string) {
+	var (
+		sampleCount uint64
+		sampleSum   float64
+		quantiles   map[float64]float64
+	)
+
+	for _, userMetrics := range d { // for each user
+		for _, m := range userMetrics[summaryName].GetMetric() {
```

Comment:
Here `userMetrics[summaryName]` could panic as well if `nil`

Reply:
Explained earlier, this isn't true.